src/components/file/media/media.repository.ts

```diff
@@ -0,0 +1,148 @@
+        node('node', 'Media'),
+      ])
+      .where(
+        or([
+          ...(input.fvIds ? [{ 'fv.id': inArray(input.fvIds) }] : []),
+          ...(input.mediaIds ? [{ 'node.id': inArray(input.mediaIds) }] : []),
+        ]),
+      )
+      .apply(this.hydrate())
+      .map('dto')
+      .run();
+  }
+
+  protected hydrate() {
+    return (query: Query) =>
+      query.return<{ dto: AnyMedia }>(
+        merge('node', {
+          __typename: 'node.type',
+          file: 'fv.id',
+          dimensions: {
+            width: 'node.width',
+            height: 'node.height',
+          },
+        }).as('dto'),
+      );
+  }
+
+  async create(input: Except<AnyMedia, 'id'>) {
+    const query = this.db
+      .query()
+      .matchNode('fv', 'FileVersion', { id: input.file })
+      .apply(
+        await createNode(resolveMedia(input), {
+          baseNodeProps: toDbShape(input),
+        }),
+      )
+      .apply(
+        createRelationships(Media, 'in', {
+          media: variable('fv'),
+        }),
+      )
+      // Grab the previous media node or null
+      .subQuery('fv', (sub) =>
+        sub
+          .optionalMatch([
+            node('fv'), // current is ignored since it's called out separately in this match
+            relation('out', '', 'parent', ACTIVE),
+            node('file', 'File'),
+            relation('in', '', 'parent', ACTIVE),
+            node('fvs', 'FileVersion'),
+          ])
+          .optionalMatch([
+            node('fvs'),
+            relation('out', '', 'media'),
+            node('prevMedia', 'Media'),
+          ])
+          .return('prevMedia')
+          .orderBy('fvs.createdAt', 'DESC')
+          .limit(1),
+      )
+      // Use previous user metadata as defaults for new media
+      .with('node, fv, prevMedia')
+      .setVariables(
+        {
+          node: String(
+            apoc.map.submap(
+              apoc.map.merge('node', 'prevMedia'),
+              MediaUserMetadata.Props.map((k) => `'${k}'`),
+              [],
+              false,
+            ),
+          ),
+        },
+        true,
+      )
+      .apply(this.hydrate());
+
+    const result = await query.first();
+    if (!result) {
+      throw new ServerException('Failed to create media info');
+    }
+    return result.dto;
+  }
+
+  async update(
+    input: RequireAtLeastOne<Pick<AnyMedia, 'id' | 'file'>> & Partial<AnyMedia>,
+  ) {
+    const query = this.db
+      .query()
+      .match([
+        node('fv', 'FileVersion', input.file ? { id: input.file } : {}),
+        relation('out', '', 'media'),
+        node('node', 'Media', input.id ? { id: input.id } : {}),
+      ])
+      .setValues({ node: toDbShape(input) }, true)
+      .apply(this.hydrate());
+
+    const result = await query.first();
+    if (!result) {
+      throw new ServerException('Failed to save media info');
+    }
+    return result.dto;
+  }
+}
+
+const toDbShape = (input: Partial<AnyMedia>) => ({
```

Comment:
I think I groked this but it was a bit hard to follow....

Probably a bit of a nit but something like this might be easier to read:
```

const getDimensions = (typename: string | undefined, dimensions: { width?: number | null, height?: number | null }) => {
  return (typename === 'Image' || typename === 'Video') 
    ? dimensions 
    : { width: null, height: null };
};

const getDuration = (typename: string | undefined, duration: number | undefined) => {
  return { duration: typename === 'Audio' ? duration : null };
};

const toDbShape = (input: Partial<AnyMedia>) => {
  return {
    type: input.__typename,
    altText: input.altText,
    caption: input.caption,
    mimeType: input.mimeType,
    ...getDimensions(input.__typename, input.dimensions),
    ...getDuration(input.__typename, input.duration),
  };
};
```

Reply:
Part of why this is syntax heavy is the nested check for null `__typename`. I wanted this to be used for multiple purposes:
1. Saving detected metadata, and if for whatever reason type changes we want to clear out previous/stale values
2. Saving user metadata. Which should just need the `altText`, `caption` as input. So we'd want to ignore typename logic there and not clear out the existing values.

That additional check could certainly be added to your helper functions above, or some combination. But IMO they don't add enough value to warrant the additional typing needed for the function inputs.